### PR TITLE
7730: gh data sourcing and screenshot generation

### LINKED ISSUE #7730: Starter Showcase: Improve data sourcing from GitHub / screenshot generation
Improve current process for sourcing GitHub data for starters.

- [x] source from GitHub directly (no scraping/committing)
- [x] ensure that sans token, front-end isn't broken for DX but fails for build
- [x] modify gatsby-transformer-screenshot to accept other yml files

### PR DESCRIPTION
**Pull request by amberleyromo**
~(not ready for merge)~ ready for review

Closes #7730

tldr; this moves the data sourcing and creation of the starter showcase landing page and starter pages in line with the other data sourcing (eg site showcase).

todos:
- [x] pull starter data from yml (in sync with site showcase flow)
- [x] update `gatsby-transformer-screenshot` to target add'l node types
- [x] update `gatsby-transformer-screenshot` to not break everything on failed max retries
- [x] convert all view code to reference yml data + new child screenshots instead of markdown/screenshot file
- [x] remove hanging references to old starter data
- [ ] handle gh token for prod
- [x] ensure lacking a gh token in dev mode doesn't break `www` DX
- [x] what is this `wat` commit? lol (dropped)

### CODE REVIEW

File: www/src/views/starter-showcase/filtered-showcase.js [hunk 1]
```diff
@@ -48,25 +48,38 @@ export default class FilteredShowcase extends Component {
       )
     )
 
-    let items = data.allMarkdownRemark.edges,
-      imgs = data.allFile.edges
+    let starters = data.allStartersYaml.edges
+
+    if (starters[0].node.fields.starterShowcase.error) {
```
Comment (amberleyromo) [inline www/src/views/starter-showcase/filtered-showcase.js]: This isn't perfect. Does need to change in some way. Would appreciate thoughts/ideas.

If there's a token, but the request for github data fails for some reason, wouldn't want this error on the built site.

Comment (amberleyromo) [reply www/src/views/starter-showcase/filtered-showcase.js]: Do you think instead of ui messaging, should just display empty values?

File: www/gatsby-node.js [hunk 2]
```diff
@@ -229,27 +261,27 @@ exports.createPages = ({ graphql, actions }) => {
         })
       })
 
-      // Create starters.
-      const starters = _.filter(result.data.allMarkdownRemark.edges, edge => {
+      // Create starter pages.
+      const starters = _.filter(result.data.allStartersYaml.edges, edge => {
         const slug = _.get(edge, `node.fields.starterShowcase.slug`)
         if (!slug) return null
         else return edge
```
[the remark this answers is not in the document]
Comment (amberleyromo) [reply www/gatsby-node.js]: Lol I can make it consistent while I'm here

File: www/src/views/starter-showcase/showcase-list.js [hunk 3]
```diff
@@ -44,56 +45,55 @@ const ShowcaseList = ({ urlState, items, imgs, count, sortRecent }) => {
       </div>
     )
   }
-  if (count) items = items.sort(sortingFunction(sortRecent)).slice(0, count)
+  if (count)
+    starters = starters.sort(sortingFunction(sortRecent)).slice(0, count)
```
[the remark this answers is not in the document]
Comment (amberleyromo) [reply www/src/views/starter-showcase/showcase-list.js]: Ah, some of this I think just got auto reformatted when I made edits. I'll add those here, no prob.

File: www/src/views/starter-showcase/filtered-showcase.js [hunk 4]
```diff
@@ -370,27 +376,26 @@ function count(arrays) {
...
+  starters = starters.filter(
+    ({ node: starter }) =>
+      starter.fields &&
       isSuperset(
-        node.fields.starterShowcase.gatsbyDependencies.map(c => c[0]),
+        starter.fields.starterShowcase.gatsbyDependencies.map(c => c[0]),
         categories
       )
-    // node.fields.starterShowcase.gatsbyDependencies.filter(c => categories.has(c[0])).length > 0
+    // starter.fields.starterShowcase.gatsbyDependencies.filter(c => categories.has(c[0])).length > 0
```
[the remark this answers is not in the document]
Comment (amberleyromo) [reply www/src/views/starter-showcase/filtered-showcase.js]: I have no idea. I touched this only for the renaming. I can just axe it.

[the remark this answers is not in the document]
Comment (jlengstorf) [reply www/src/views/starter-showcase/filtered-showcase.js]: Yeah, if you don't know what it's for, let's kill it. I'm a big fan of axing commented-out code unless I intend to come back to it within a day or so. In my experience, any longer than that and it just becomes deleted code that I never bothered to delete.

File: www/src/views/starter-showcase/showcase-list.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (NickyMeuleman) [reply www/src/views/starter-showcase/showcase-list.js]: Yup, got autoformatted. Prettier is now ran on staged files during pre-commit. The existing files weren't ran through Prettier first, so expect a lot of this 😅 https://github.com/gatsbyjs/gatsby/pull/8229

[the remark this answers is not in the document]
Comment (jlengstorf) [reply www/src/views/starter-showcase/showcase-list.js]: Thanks @NickyMeuleman! I just commented on that PR. Thanks for putting that in place!

File: www/gatsby-node.js [hunk 5]
```diff
@@ -401,6 +426,108 @@ exports.onCreateNode = ({ node, actions, getNode, getNodes }) => {
     const cleaned = parsed.hostname + parsed.pathname
     slug = `/showcase/${slugify(cleaned)}`
     createNodeField({ node, name: `slug`, value: slug })
+  } else if (node.internal.type === `StartersYaml` && node.repo) {
+    // To develop on the starter showcase, you'll need a GitHub
+    // personal access token. Check the `www` README for details.
+    // Default fields are to avoid graphql errors.
+    const [owner, repoStub] = node.repo.split(`/`).splice(-2, 2)
```
[the remark this answers is not in the document]
Comment (amberleyromo) [reply www/gatsby-node.js]: Decided to use the pkg 👌

File: www/src/views/starter-showcase/filtered-showcase.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (amberleyromo) [reply www/src/views/starter-showcase/filtered-showcase.js]: Axed

File: www/src/views/starter-showcase/filtered-showcase.js (hunk 1 above)
Comment (amberleyromo) [reply www/src/views/starter-showcase/filtered-showcase.js]: Done

File: www/src/views/starter-showcase/showcase-list.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (amberleyromo) [reply www/src/views/starter-showcase/showcase-list.js]: Added back braces
